silverback/persistence.py

```diff
@@ -0,0 +1,327 @@
+import json
+import os
+import sqlite3
+from abc import ABC, abstractmethod
+from datetime import datetime, timezone
+from typing import Optional, TypeVar
+
+from pydantic import BaseModel
+from taskiq import TaskiqResult
+from typing_extensions import Self  # Introduced 3.11
+
+from .types import IntOrNone, SilverbackIdent
+
+_HandlerReturnType = TypeVar("_HandlerReturnType")
+
+
+class SilverbackState(BaseModel):
+    instance: str
+    network: str
+    # Last block number seen by runner
+    last_block_seen: int
+    # Last block number processed by a worker
+    last_block_processed: int
+    updated: datetime
+
+
+class HandlerResult(TaskiqResult):
+    instance: str
+    network: str
+    handler_id: str
+    block_number: Optional[int]
+    log_index: Optional[int]
+    created: datetime
+
+    @classmethod
+    def from_taskiq(
+        cls,
+        ident: SilverbackIdent,
+        handler_id: str,
+        block_number: IntOrNone,
+        log_index: IntOrNone,
+        result: TaskiqResult,
+    ) -> Self:
+        return cls(
+            instance=ident.identifier,
+            network=ident.network_choice,
+            handler_id=handler_id,
+            block_number=block_number,
+            log_index=log_index,
+            created=datetime.now(timezone.utc),
+            **result.dict(),
+        )
+
+
+class BasePersistentStorage(ABC):
+    @abstractmethod
+    async def init(self):
+        """Handle any async initialization from Silverback settings (e.g. migrations)."""
+        ...
+
+    @abstractmethod
+    async def get_instance_state(self, ident: SilverbackIdent) -> Optional[SilverbackState]:
```

Comment:
is `instance` necessary to disambiguate?

Reply:
Do you mean the attribute `SilverbackState.instance`?  That allows multiple Silverback instances to leverage the same DB.

Reply:
> Do you mean the attribute `SilverbackState.instance`? That allows multiple Silverback instances to leverage the same DB.

no just having `instance` in this method name

Reply:
oh, gotcha.  Will shorten.